### PR TITLE
Issue #2231, Refactored my_key_update file

### PR DESCRIPTION
**Pull request by michael-volynets**
Issue #2231,
Time Spent: 40m

### CODE REVIEW

File: extension/chrome/settings/modules/my_key_update.ts [hunk 1]
```diff
@@ -6,79 +6,96 @@ import { Catch } from '../../../js/common/platform/catch.js';
...
+  private readonly prvHeaders = Pgp.armor.headers('privateKey');
 
-  const uncheckedUrlParams = Url.parse(['acctEmail', 'longid', 'parentTabId']);
-  const acctEmail = Assert.urlParamRequire.string(uncheckedUrlParams, 'acctEmail');
-  const longid = Assert.urlParamRequire.optionalString(uncheckedUrlParams, 'longid') || 'primary';
-  const showKeyUrl = Url.create('my_key.htm', uncheckedUrlParams);
+  constructor() {
+    super();
+    const uncheckedUrlParams = Url.parse(['acctEmail', 'longid', 'parentTabId']);
+    const acctEmail = Assert.urlParamRequire.string(uncheckedUrlParams, 'acctEmail');
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key_update.ts]: Here you can directly assign to `this.acctEmail`

File: extension/chrome/settings/modules/my_key_update.ts [hunk 2]
```diff
@@ -6,79 +6,96 @@ import { Catch } from '../../../js/common/platform/catch.js';
...
+    const longid = Assert.urlParamRequire.optionalString(uncheckedUrlParams, 'longid') || 'primary';
+    this.acctEmail = acctEmail;
+    this.showKeyUrl = Url.create('my_key.htm', uncheckedUrlParams);
+    this.primariKiPromise = (async () => (await Store.keysGet(acctEmail, [longid]))[0])();
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key_update.ts]: Just put this into `render`, for simplicity

File: extension/chrome/settings/modules/my_key_update.ts [hunk 3]
```diff
@@ -6,79 +6,96 @@ import { Catch } from '../../../js/common/platform/catch.js';
...
+  }
 
-  $('.action_show_public_key').attr('href', showKeyUrl);
-  const inputPrivateKey = $('.input_private_key');
-  const prvHeaders = Pgp.armor.headers('privateKey');
+  async render() {
+    const primaryKi = await this.primariKiPromise;
+    $('.action_show_public_key').attr('href', this.showKeyUrl);
+    Assert.abortAndRenderErrorIfKeyinfoEmpty(primaryKi);
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key_update.ts]: This should be directly below `[primaryKi]`, it's related

File: extension/chrome/settings/modules/my_key_update.ts [hunk 4]
```diff
@@ -6,79 +6,96 @@ import { Catch } from '../../../js/common/platform/catch.js';
...
+    $('.email').text(this.acctEmail);
+    $('.key_words').text(primaryKi.keywords).attr('title', primaryKi.longid);
+    this.inputPrivateKey.attr('placeholder', this.inputPrivateKey.attr('placeholder') + ' (' + primaryKi.longid + ')');
+  }
 
-  const [primaryKi] = await Store.keysGet(acctEmail, [longid]);
-
-  Assert.abortAndRenderErrorIfKeyinfoEmpty(primaryKi);
+  setHandlers() {
+    $('.action_update_private_key').click(this.setHandlerPrevent('double', this.updatePrivateKeyHandler));
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key_update.ts]: Clarification - all methods should still be standard class methods, but we will always be adding `el => this.updatePrivateKeyHandler()` instead of just `this.updatePrivateKeyHandler`. That's what I meant by saying we'll use an arrow function.

All methods will be standard class methods though.

File: extension/chrome/settings/modules/my_key_update.ts [hunk 5]
```diff
@@ -6,79 +6,96 @@ import { Catch } from '../../../js/common/platform/catch.js';
...
+        await Ui.modal.error(`Error updating public records:\n\n${Api.err.eli5(e)}\n\n(but local update was successful)`);
+      }
+    }
+    window.location.href = this.showKeyUrl;
+  }
 
-  $('.action_update_private_key').click(Ui.event.prevent('double', async () => {
-    const { keys: [uddatedKey] } = await openpgp.key.readArmored(String(inputPrivateKey.val()));
-    const { keys: [uddatedKeyEncrypted] } = await openpgp.key.readArmored(String(inputPrivateKey.val()));
+  private updatePrivateKeyHandler = async () => {
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key_update.ts]: Please write as standard class method, as mentioned above

File: extension/chrome/settings/modules/my_key_update.ts [hunk 6]
```diff
@@ -6,79 +6,96 @@ import { Catch } from '../../../js/common/platform/catch.js';
...
+    const primaryKi = await this.primariKiPromise;
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key_update.ts]: Same here, don't work with promises, just get the variable in `render` which is async, then directly access variable.

File: extension/chrome/settings/modules/my_key_update.ts [hunk 7]
```diff
@@ -6,79 +6,96 @@ import { Catch } from '../../../js/common/platform/catch.js';
...
-      await Ui.modal.warning('This was a public key. Please insert a private key instead. It\'s a block of text starting with "' + prvHeaders.begin + '"');
+      await Ui.modal.warning('This was a public key. Please insert a private key instead. It\'s a block of text starting with "' + this.prvHeaders.begin + '"');
     } else if (await Pgp.key.fingerprint(uddatedKey) !== await Pgp.key.fingerprint(primaryKi.public)) {
       await Ui.modal.warning(`This key ${await Pgp.key.longid(uddatedKey)} does not match your current key ${primaryKi.longid}`);
     } else if (await Pgp.key.decrypt(uddatedKey, uddatedKeyPassphrase) !== true) {
       await Ui.modal.error('The pass phrase does not match.\n\nPlease enter pass phrase of the newly updated key.');
     } else {
       if (await uddatedKey.getEncryptionKey()) {
-        await storeUpdatedKeyAndPassphrase(uddatedKeyEncrypted, uddatedKeyPassphrase);
+        await this.storeUpdatedKeyAndPassphrase(primaryKi, uddatedKeyEncrypted, uddatedKeyPassphrase);
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key_update.ts]: use `this.primaryKi!`

File: extension/chrome/settings/modules/my_key_update.ts [hunk 8]
```diff
@@ -6,79 +6,96 @@ import { Catch } from '../../../js/common/platform/catch.js';
...
       } else { // cannot get a valid encryption key packet
         if ((await uddatedKey.verifyPrimaryKey() === openpgp.enums.keyStatus.no_self_cert) || await Pgp.key.usableButExpired(uddatedKey)) { // known issues - key can be fixed
           const fixedEncryptedPrv = await Settings.renderPrvCompatFixUiAndWaitTilSubmittedByUser(
-            acctEmail, '.compatibility_fix_container', uddatedKeyEncrypted, uddatedKeyPassphrase, showKeyUrl
+            this.acctEmail, '.compatibility_fix_container', uddatedKeyEncrypted, uddatedKeyPassphrase, this.showKeyUrl
           );
-          await storeUpdatedKeyAndPassphrase(fixedEncryptedPrv, uddatedKeyPassphrase);
+          await this.storeUpdatedKeyAndPassphrase(primaryKi, fixedEncryptedPrv, uddatedKeyPassphrase);
```
Comment (tomholub) [inline extension/chrome/settings/modules/my_key_update.ts]: use `this.primaryKi!`

File: extension/chrome/settings/modules/my_key_update.ts (hunk 2 above)
Comment (michael-volynets) [reply extension/chrome/settings/modules/my_key_update.ts]: I also was thinking about it but if I put it into `render` method the type of `primariKi` will be `KeyInfo | undefined`

File: extension/chrome/settings/modules/my_key_update.ts (hunk 3 above)
Comment (michael-volynets) [reply extension/chrome/settings/modules/my_key_update.ts]: Ok, Will move

File: extension/chrome/settings/modules/my_key_update.ts (hunk 4 above)
Comment (michael-volynets) [reply extension/chrome/settings/modules/my_key_update.ts]: Got it, Will change it

File: extension/chrome/settings/modules/my_key_update.ts (hunk 2 above)
Comment (tomholub) [reply extension/chrome/settings/modules/my_key_update.ts]: I'd force it with `!`, I think that's a cleaner solution, although not perfect.
